jgrapht-opt/src/main/java/org/jgrapht/opt/graph/sparse/DefaultSparseGraph.java

```diff
@@ -0,0 +1,555 @@
+    {
+        assertVertexExist(vertex);
+        return edgeStream(vertex, true, directed ? false : true)
+            .boxed().collect(Collectors.toSet());
+    }
+
+    @Override
+    public Integer removeEdge(Integer sourceVertex, Integer targetVertex)
+    {
+        throw new UnsupportedOperationException(UNMODIFIABLE);
+    }
+
+    @Override
+    public boolean removeEdge(Integer e)
+    {
+        throw new UnsupportedOperationException(UNMODIFIABLE);
+    }
+
+    @Override
+    public boolean removeVertex(Integer v)
+    {
+        throw new UnsupportedOperationException(UNMODIFIABLE);
+    }
+
+    @Override
+    public Set<Integer> vertexSet()
+    {
+        return new IntegerSet(n);
+    }
+
+    @Override
+    public Integer getEdgeSource(Integer e)
+    {
+        assertEdgeExist(e);
+        if (directed || source[e] <= target[e]) {
+            return source[e];
+        } else {
+            return target[e];
+        }
+    }
+
+    @Override
+    public Integer getEdgeTarget(Integer e)
+    {
+        assertEdgeExist(e);
+        if (directed || source[e] <= target[e]) {
+            return target[e];
+        } else {
+            return source[e];
+        }
+    }
+
+    @Override
+    public GraphType getType()
+    {
+        Builder builder = new DefaultGraphType.Builder();
+        builder = (directed) ? builder.directed() : builder.undirected();
+        return builder
+            .weighted(false).modifiable(false).allowMultipleEdges(true).allowSelfLoops(true)
+            .build();
+    }
+
+    @Override
+    public double getEdgeWeight(Integer e)
+    {
+        return Graph.DEFAULT_EDGE_WEIGHT;
+    }
+
+    @Override
+    public void setEdgeWeight(Integer e, double weight)
+    {
+        throw new UnsupportedOperationException(UNMODIFIABLE);
+    }
+
+    /**
+     * Ensures that the specified vertex exists in this graph, or else throws exception.
+     *
+     * @param v vertex
+     * @return <code>true</code> if this assertion holds.
+     * @throws IllegalArgumentException if specified vertex does not exist in this graph.
+     */
+    protected boolean assertVertexExist(Integer v)
+    {
+        if (v >= 0 && v < n) {
+            return true;
+        } else {
+            throw new IllegalArgumentException("no such vertex in graph: " + v.toString());
+        }
+    }
+
+    /**
+     * Ensures that the specified edge exists in this graph, or else throws exception.
+     *
+     * @param e edge
+     * @return <code>true</code> if this assertion holds.
+     * @throws IllegalArgumentException if specified edge does not exist in this graph.
+     */
+    protected boolean assertEdgeExist(Integer e)
+    {
+        if (e >= 0 && e < m) {
+            return true;
+        } else {
+            throw new IllegalArgumentException("no such edge in graph: " + e.toString());
+        }
+    }
+
+    /**
+     * A comparator for indirect sort.
+     */
+    private class IndirectComparator
```

Comment:
This inner class should be static.  The same goes for a bunch of other inner classes in this PR.